akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -354,4 +395,57 @@ object HttpEntity {
     def getTrailerHeaders: JIterable[jm.HttpHeader] = trailer.asJava
   }
   object LastChunk extends LastChunk("", Nil)
+
+  /**
+   * Turns the given source into one that respects the `withSizeLimit` calls when used as a parameter
+   * to entity constructors.
+   */
+  def limitableByteSource[Mat](source: Source[ByteString, Mat]): Source[ByteString, Mat] =
+    limitable(source, sizeOfByteString)
+
+  /**
+   * Turns the given source into one that respects the `withSizeLimit` calls when used as a parameter
+   * to entity constructors.
+   */
+  def limitableChunkSource[Mat](source: Source[ChunkStreamPart, Mat]): Source[ChunkStreamPart, Mat] =
+    limitable(source, sizeOfChunkStreamPart)
+
+  /**
+   * INTERNAL API
+   */
+  private val sizeOfByteString: ByteString ⇒ Int = _.size
+  private val sizeOfChunkStreamPart: ChunkStreamPart ⇒ Int = _.data.size
+
+  /**
+   * INTERNAL API
+   */
+  private def limitable[Out, Mat](source: Source[Out, Mat], sizeOf: Out ⇒ Int): Source[Out, Mat] =
+    source.via(Flow[Out].transform { () ⇒
+      new PushStage[Out, Out] {
+        var maxBytes = -1L
+        var bytesLeft = Long.MaxValue
+
+        override def preStart(ctx: LifecycleContext) =
+          ctx.attributes.getFirst[SizeLimit] match {
+            case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
+              if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
+            // else we still count but never throw an error
+            case Some(SizeLimit(bytes, None)) ⇒
+              maxBytes = bytes
+              bytesLeft = bytes
+            case None ⇒
+          }
+
+        def onPush(elem: Out, ctx: stage.Context[Out]): stage.SyncDirective = {
+          bytesLeft -= sizeOf(elem)
+          if (bytesLeft >= 0) ctx.push(elem)
+          else ctx.fail(EntityStreamSizeException(maxBytes))
+        }
+      }
+    }.named("limit"))
```

Comment:
Though technically `limitable`; The reason I'm picky here is because we're considering a `limit` op on Flow ops ( https://github.com/akka/akka/issues/18921 ), so the name "limit" used here would be a bit misleading (better fix now, than when adding limit IMHO)